Fall back to IPv4 if IPv6 connection fails

### Expected behaviour

POP mail should download on timed intervals or on demand
### Actual behaviour

POP mail does not download - fails immediately. Mail can still be sent and other Internet applications work.
### Steps to reproduce

Problem started only after I switched ISP. Wireless router unchanged. Everything works perfectly until I am out of range of my home WiFi network. F-9 continues to work perfectly using the phone network.

When I return to my home WiFi network, the WiFi network re-connects and it is used preferentially over the phone network, but K-9 no longer fetches mail. K-9 can still send (it uses a different server than for fetching which may or may be relevant) and all other Internet applications are fine.

Neither a Force Stop of K-9 nor an Android reboot fixes the problem. Only solution I have found is to temporarily reconnect my old modem/ISP (which is still working for a few days) and reconnect to the POP server. This resets the problem and it now works when I revert to the new ISP (until I go out of WiFi range again).

Debug logs implies that there is a 30000 MS timeout connecting to port 995, but the error happens almost instantaneously. See log for two consecutive tries 2 seconds apart.

 Debug Logs (3 instances - on cursory inspection, they seem to all be the same) attached
[DebugLog-20160514.txt](https://github.com/k9mail/k-9/files/264781/DebugLog-20160514.txt)
### Environment

K-9 Mail version: 5.010
Android version: 5.1.1
Account type (IMAP, POP3, WebDAV/Exchange): POP3

I see the issue for both POP3 and SMTP as well (my mail server has IPv4 and IPv6 support) - at hotel wlan or similar low-cost setups over which I have no controll IPv6 might be available but not work, or routing is temporarily broken.
I have to turn off wlan and use LTE to get messages out.

K9-Mail version: 5.206

com.fsck.k9.mail.MessagingException: Unable to open connection to SMTP server.
	at com.fsck.k9.mail.transport.SmtpTransport.open(SmtpTransport.java:405)
	at com.fsck.k9.mail.transport.SmtpTransport.sendMessageTo(SmtpTransport.java:504)
	at com.fsck.k9.mail.transport.SmtpTransport.sendMessage(SmtpTransport.java:496)
	at com.fsck.k9.controller.MessagingController.sendPendingMessagesSynchronous(MessagingController.java:3014)
	at com.fsck.k9.controller.MessagingController.access$1700(MessagingController.java:115)
	at com.fsck.k9.controller.MessagingController$21.run(MessagingController.java:2895)
	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.net.SocketTimeoutException: failed to connect to mailserver.example.com/an:IPv6::address (port 25) after 10000ms
	at libcore.io.IoBridge.connectErrno(IoBridge.java:169)
	at libcore.io.IoBridge.connect(IoBridge.java:122)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:183)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:452)
	at java.net.Socket.connect(Socket.java:884)
	at com.fsck.k9.mail.transport.SmtpTransport.open(SmtpTransport.java:223)

This is a BIG problem. I cannot send mail to any provider offering IPV6 on their SMTP server - not even my own. The phone picks a private IPV6 address, the server offers one but none of my routers and none of my internet providers offer an IPV6 path.

07-12 23:25:29.670 E/k9      ( 5752): Caused by: java.net.SocketTimeoutException: failed to connect to mail.stellarhost.eu/2a01:4f8:190:2461::200:1 (port 587) after 10000ms

The same to gmail.

None of any other mail clients has any problem with this and READING mail seems no problem, even if the servers offer IPV6 on IMAP too - which they do.

PLEASE, provide a fallback and please keep the timeout reasonable.

--- a/mail/protocols/pop3/src/main/java/com/fsck/k9/mail/store/pop3/Pop3Connection.java
+++ b/mail/protocols/pop3/src/main/java/com/fsck/k9/mail/store/pop3/Pop3Connection.java
@@ -5,9 +5,10 @@ import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.Socket;
-import java.net.SocketAddress;
+import java.net.UnknownHostException;
 import java.security.GeneralSecurityException;
 import java.security.KeyManagementException;
 import java.security.MessageDigest;
@@ -61,15 +62,7 @@ class Pop3Connection {
 
     void open() throws MessagingException {
         try {
-            SocketAddress socketAddress = new InetSocketAddress(settings.getHost(), settings.getPort());
-            if (settings.getConnectionSecurity() == ConnectionSecurity.SSL_TLS_REQUIRED) {
-                socket = trustedSocketFactory.createSocket(null, settings.getHost(),
-                        settings.getPort(), settings.getClientCertificateAlias());
-            } else {
-                socket = new Socket();
-            }
-
-            socket.connect(socketAddress, SOCKET_CONNECT_TIMEOUT);
+            socket = connect();
             in = new BufferedInputStream(socket.getInputStream(), 1024);
             out = new BufferedOutputStream(socket.getOutputStream(), 512);
 
@@ -100,6 +93,44 @@ class Pop3Connection {
         } catch (IOException ioe) {
             throw new MessagingException("Unable to open connection to POP server.", ioe);
         }
+    }
+
+    private Socket connect()
+            throws IOException, MessagingException, NoSuchAlgorithmException, KeyManagementException {
+        InetAddress[] inetAddresses = InetAddress.getAllByName(settings.getHost());
+
+        IOException connectException = null;
+        for (InetAddress address : inetAddresses) {
+            try {
+                return connectToAddress(address);
+            } catch (IOException e) {
+                Timber.w(e, "Could not connect to %s", address);
+                connectException = e;
+            }
+        }
+
+        throw connectException != null ? connectException : new UnknownHostException();
+    }
+
+    private Socket connectToAddress(InetAddress address)
+            throws IOException, MessagingException, NoSuchAlgorithmException, KeyManagementException {
+        if (K9MailLib.isDebug() && K9MailLib.DEBUG_PROTOCOL_POP3) {
+            Timber.d("Connecting to %s as %s", settings.getHost(), address);
+        }
+
+        InetSocketAddress socketAddress = new InetSocketAddress(address, settings.getPort());
+
+        final Socket socket;
+        if (settings.getConnectionSecurity() == ConnectionSecurity.SSL_TLS_REQUIRED) {
+            socket = trustedSocketFactory.createSocket(null, settings.getHost(), settings.getPort(),
+                    settings.getClientCertificateAlias());
+        } else {
+            socket = new Socket();
+        }
+
+        socket.connect(socketAddress, SOCKET_CONNECT_TIMEOUT);
+
+        return socket;
     }
 
     /*

--- a/mail/protocols/pop3/src/test/java/com/fsck/k9/mail/store/pop3/Pop3StoreTest.kt
+++ b/mail/protocols/pop3/src/test/java/com/fsck/k9/mail/store/pop3/Pop3StoreTest.kt
@@ -37,7 +37,7 @@ class Pop3StoreTest {
     @Test(expected = MessagingException::class)
     fun `checkSettings() with TrustedSocketFactory throwing should throw MessagingException`() {
         stubbing(trustedSocketFactory) {
-            on { createSocket(null, "server", 12345, null) } doThrow IOException()
+            on { createSocket(null, HOST, 12345, null) } doThrow IOException()
         }
 
         store.checkSettings()
@@ -72,7 +72,7 @@ class Pop3StoreTest {
     private fun createServerSettings(): ServerSettings {
         return ServerSettings(
             type = "pop3",
-            host = "server",
+            host = HOST,
             port = 12345,
             connectionSecurity = ConnectionSecurity.SSL_TLS_REQUIRED,
             authenticationType = AuthType.PLAIN,
@@ -93,13 +93,14 @@ class Pop3StoreTest {
         }
 
         stubbing(trustedSocketFactory) {
-            on { createSocket(null, "server", 12345, null) } doReturn socket
+            on { createSocket(null, HOST, 12345, null) } doReturn socket
         }
 
         return outputStream
     }
 
     companion object {
+        private const val HOST = "127.0.0.1"
         private const val INITIAL_RESPONSE = "+OK POP3 server greeting\r\n"
 
         private const val CAPA_RESPONSE = "+OK Listing of supported mechanisms follows\r\n" +

